src/main/scala/loamstream/model/execute/RxExecuter.scala

```diff
@@ -54,58 +55,111 @@ final case class RxExecuter(
+      ExecutionState.initialFor(executable, maxRunsPerJob)
+    }
+
+    val ioScheduler: Scheduler = IOScheduler()
     
-    val chunkResults: Observable[Map[LJob, Execution]] = for {
-      //NB: .toSet is important: jobs in a chunk should be distinct, 
-      //so they're not run more than once before transitioning to a terminal state.
-      jobs <- chunks.map(_.toSet)
-      //NB: Filter out jobs from this chunk that finished when run as part of another chunk, so we don't run them
-      //more times than necessary.  This helps in the face of job-restarting, since we can't call `distinct()` 
-      //on `runnables` and declare victory like we did before, since that would filter out restarting jobs that 
-      //already ran 
-      (finishedJobs, notFinishedJobs) = jobs.partition(_.status.isTerminal)
-      if notFinishedJobs.nonEmpty
-      (jobsToMaybeRun, skippedJobs) = notFinishedJobs.map(_.job).partition(jobFilter.shouldRun)
-      (jobsToCancel, jobsToRun) = jobsToMaybeRun.partition(jobCanceler.shouldCancel)
-      _ = handleSkippedJobs(skippedJobs)
-      cancelledJobsMap = cancelJobs(jobsToCancel)
-      _ = record(jobOracle, cancelledJobsMap)
-      skippedResultMap = toSkippedResultMap(skippedJobs)
-      executionTupleOpt <- runJobs(jobsToRun, jobOracle)
-      _ = record(jobOracle, executionTupleOpt)
-      executionMap = cancelledJobsMap ++ executionTupleOpt
-      _ = logFinishedJobs(executionMap)
-    } yield {
-      executionMap ++ skippedResultMap
+    val chunkResults: Observable[Map[LJob, Execution]] = {
+      //Note onBackpressureDrop(), in case runEligibleJobs takes too long (or the polling window is too short)
+      val ticks = Observable.interval(windowLength, ioScheduler).onBackpressureDrop
+      
+      def runJobs() = runEligibleJobs(executionState, jobOracle)
+      
+      def isFinished = executionState.isFinished
+      
+      ticks.flatMap(_ => runJobs).takeUntil(_ => isFinished)
     }
-    //NB: We no longer stop on the first failure, but run each sub-tree of jobs as far as possible.
     
     val futureMergedResults = chunkResults.foldLeft(emptyExecutionMap)(_ ++ _).firstAsFuture
 
     Await.result(futureMergedResults, timeout)
   }
   
+  private def finish(executionState: ExecutionState)(results: Map[LJob, Execution]): Map[LJob, Execution] = {
+    def msg = {
+      val howMany: Int = scala.math.min(50, results.size) //scalastyle:ignore magic.number
+      
+      s"Finishing ${results.size} jobs; first $howMany ids: ${results.keys.take(howMany).map(_.id).mkString(",")}"
+    }
+    
+    TimeUtils.time(msg, debug(_)) {
+      executionState.finish(results)
+    }
+    
+    results
+  }
+  
+  private def runEligibleJobs(
+      executionState: ExecutionState, 
+      jobOracle: JobOracle): Observable[Map[LJob, Execution]] = {
+      
+    val jobsAndCells = executionState.updateJobs()
+    
+    val (numReadyToRun, numCannotRun) = (jobsAndCells.readyToRun.size, jobsAndCells.cannotRun.size)
```

Comment:
Is `numCannotRun` a synonym for `isNotReadyToRun`? Just want to understand if it is potentially a state of "tried to run and failed, so assuming it can't, but is different from finished".

Reply:
`numReadyToRun` is the number of jobs that LS knows it _can_ currently run.  `numCannotRun` is the number of jobs that LS knows it _cannot_ currently run.  

There may be other jobs that LS doesn't know what to do with yet, like ones that depend on a runnable job; they may end up running if the job they depend on succeeds, or they may not if it fails.

There are also jobs that have failed but are eligible to be re-run, and some other edge cases.